stdlib/result.mli

```diff
@@ -0,0 +1,103 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*   Copyright 2018 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Result values.
+
+    Result values handle computation results and errors in an explicit
+    and declarative manner without resorting to exceptions.
+
+    @since 4.08 *)
+
+(** {1:results Results} *)
+
+type ('a, 'e) t = ('a, 'e) result = Ok of 'a | Error of 'e
+(** The type for result values. Either a value [Ok v] or an error [Error e]. *)
+
+val ok : 'a -> ('a, 'e) result
+(** [ok v] is [Ok v]. *)
+
+val error : 'e -> ('a, 'e) result
+(** [error e] is [Error e]. *)
+
+val value : ('a, 'e) result -> default:'a -> 'a
+(** [value r ~default] is [v] if [r] is [Ok v] and [default] otherwise. *)
+
+val get_ok : ('a, 'e) result -> 'a
+(** [get_ok r] is [v] if [r] is [Ok v] and @raise Invalid_argument
+    otherwise. *)
+
+val get_error : ('a, 'e) result -> 'e
+(** [get_error r] is [e] if [r] is [Error e] and @raise Invalid_argument
+    otherwise. *)
+
+val bind : ('a, 'e) result -> ('a -> ('b, 'e) result) -> ('b, 'e) result
+(** [bind r f] is [Ok (f v)] if [r] is [Ok v] and [r] if [r] is [Error _]. *)
+
+val join : (('a, 'e) result, 'e) result -> ('a, 'e) result
+(** [join rr] is [r] if [rr] is [Ok r] and [rr] if [rr] is [Error _]. *)
+
+val map : ('a -> 'b) -> ('a, 'e) result -> ('b, 'e) result
+(** [map f r] is [Ok (f v)] if [r] is [Ok v] and [r] if [r] is [Error _]. *)
+
+val map_error : ('e -> 'f) -> ('a, 'e) result -> ('a, 'f) result
+(** [map_error f r] is [Error (f e)] if [r] is [Error e] and [r] if
+    [r] is [Ok _]. *)
+
+val fold : ('a -> 'c) -> ('e -> 'c) -> ('a, 'e) result -> 'c
```

Comment:
If `equal` and `compare` take labelled `~ok` and `~error` parameters, why not `fold`?

Reply:
`fold` is likely to be partially applied and when you partially apply the labels leak in the result which is often annoying. For `compare` and `equal` it is much less likely that you'll partially apply them.

Reply:
(For an example of this see @pqwy's [examples](https://github.com/ocaml/ocaml/pull/1940#issuecomment-408938708) for `fold` (`bacon` sic) on the `Option` module, had the argument been labelled those couldn't be written without eta expanding).

Reply:
Eta expanding is often a good thing though, in my experience.  I've found it means less work when refactoring.  Clever partial application tricks should be avoided as much as possible, IMO.

That said, if you're going to partially apply only one of [ok] or [error], isn't it going to be clearer at the place where the other of those is finally applied to have a label?  I think it probably is, if that place is some distance away.

Labelling them also means that the argument of type "result" can be written first.  This keeps the thing being folded over close to the word "fold" in the code, rather than potentially many lines lower, resulting in easier-to-read code.

Reply:
Ok I'd like to bring all this to an end so I labelled both `{Result,Option}.fold`